lib/Mojo/WebSocketProxy/Backend/ConsumerGroups.pm

```diff
@@ -0,0 +1,345 @@
+
+    my $rpc_response_cb = $self->get_rpc_response_cb($c, $req_storage);
+    my $before_get_rpc_response_hooks = delete($req_storage->{before_get_rpc_response}) || [];
+    my $after_got_rpc_response_hooks  = delete($req_storage->{after_got_rpc_response})  || [];
+    my $before_call_hooks             = delete($req_storage->{before_call})             || [];
+    my $rpc_failure_cb                = delete($req_storage->{rpc_failure_cb});
+
+    foreach my $hook ($before_call_hooks->@*) { $hook->($c, $req_storage) }
+
+    $self->request($request_data)->then(
+        sub {
+            my ($message) = @_;
+
+            foreach my $hook ($before_get_rpc_response_hooks->@*) { $hook->($c, $req_storage) }
+
+            return Future->done unless $c && $c->tx;
+            my $api_response;
+            my $result;
+
+            try {
+                $result = MojoX::JSON::RPC::Client::ReturnObject->new(rpc_response => $message);
+                foreach my $hook ($after_got_rpc_response_hooks->@*) { $hook->($c, $req_storage, $result) }
+                $api_response = $rpc_response_cb->($result->result);
+            }
+            catch {
+                my $error = $@;
+                $rpc_failure_cb->($c, $result, $req_storage) if $rpc_failure_cb;
+                $api_response = $c->wsp_error($msg_type, 'WrongResponse', 'Sorry, an error occurred while processing your request.');
+            };
+
+            $c->send({json => $api_response}, $req_storage);
+
+            return Future->done;
+        }
+        )->catch(
+        sub {
+            my $error = shift;
+            my $api_response;
+
+            return Future->done unless $c && $c->tx;
+
+            if ($error eq 'Timeout') {
+                $api_response = $c->wsp_error($msg_type, 'RequestTimeout', 'Request is timed out.');
+            } else {
+                $api_response = $c->wsp_error($msg_type, 'WrongResponse', 'Sorry, an error occurred while processing your request.');
+            }
+            $rpc_failure_cb->($c, undef, $req_storage) if $rpc_failure_cb;
+
+            $c->send({json => $api_response}, $req_storage);
+        })->retain;
+
+    return;
+}
+
+=head2 request
+
+Sends request to backend service. The method accepts single unnamed argument:
+
+=over 4
+
+=item * C<request_data> - an C<arrayref> containing data for the item which is going to be put into redis stream.
+
+=back
+
+Returns future.
+Which will be marked as done in case getting response from backend server.
+And it'll be marked as failed in case of request timeout or in case of error putting request to redis stream.
+
+=cut
+
+sub request {
+    my ($self, $request_data) = @_;
+
+    my $complete_future = $self->loop->new_future;
+
+    $self->wait_for_messages();
+
+    my $sent_future = $self->_send_request($request_data)->then(
+        sub {
+            my ($msg_id) = @_;
+
+            $self->pending_requests->{$msg_id} = $complete_future;
```

Comment:
This looks like a race condition - I'd expect to be setting this *before* we send out the message.

Reply:
By specification we should use id, that we get from redis.

Reply:
probably you right, there is a room for race condition. Let me check with @ea-binary  about it.

Reply:
@kirill-binary yes, let's change the specs.